src/nvim/api/private/helpers.c

```diff
@@ -936,12 +936,33 @@ bool api_dict_to_keydict(void *retval, FieldHashfn hashy, Dictionary dict, Error
 
     char *mem = ((char *)retval + field->ptr_off);
     Object *value = &dict.items[i].value;
+
     if (field->type == kObjectTypeNil) {
       *(Object *)mem = *value;
     } else if (field->type == kObjectTypeInteger) {
+      if (field->is_hlgroup) {
+        if (value->type != kObjectTypeNil && value->type != kObjectTypeString
+            && value->type != kObjectTypeInteger) {
+          api_set_error(err, kErrorTypeValidation, "Invalid type of '%.*s'", (int)k.size, k.data);
+          return false;
+        }
+
+        if (value->type == kObjectTypeNil) {
```

Comment:
after checking for nil, we can use `object_to_hl_id()` for both string, int handling, and also type error checking (`what` should be `k.data` )